jax_unirep/evotuning.py

```diff
@@ -98,23 +99,75 @@ def batch_iter(
     for xmat, ymat in zip(xs, ys):
         # Send x and y in small batches of 100 to control memory usage.
         for x, y in batch_iter(xmat, ymat, batch_size=batch_size):
-            sum_loss += evotune_loss_jit(params, inputs=x, targets=y) * len(x)
+            sum_loss += evotune_loss_jit(
+                params=params, inputs=x, targets=y
+            ) * len(x)
             num_seqs += len(x)
 
     return sum_loss / num_seqs
 
 
+def generate_one_length_batch(
+    sequences: Iterable[str], holdout_seqs: Optional[Iterable[str]] = None
+) -> Tuple[int, Iterable[str], Optional[Iterable[str]]]:
+    """
+    Generates a single-length batch.
+
+    This function is refactored out of the ``fit`` function
+    to make it easier to read.
+
+    :param sequences: Sequences to generate one length batch for.
+    :param holdout_seqs: Holdout sequences.
+    """
+    # First pad to the same length, effectively giving us one length batch.
+    all_sequences = set(sequences)
+    if holdout_seqs is not None:
+        all_sequences = all_sequences.union(set(holdout_seqs))
+    max_len = max([len(seq) for seq in all_sequences])
+    sequences = right_pad(sequences, max_len)
+    if holdout_seqs is not None:
+        holdout_seqs = right_pad(holdout_seqs, max_len)
+    return max_len, sequences, holdout_seqs
+
+
+def generate_batching_funcs(
+    sequences: Iterable[str], batch_size: int
+) -> Tuple[Dict[int, Callable], List[List[str]], List[int]]:
+    """
+    Generate a batching function for each sequence length
+
+    Given a set of sequences and a batch size,
+    this function generates a dictionary,
+    where each key value pair consists of a
+    unique sequence length and a batching function for that length
+    respectively.
+    It also returns the batched sequences,
+    as well as the unique sequence lenghts.
+
+    :param sequences: Sequences to generate batching functions for
+    :param batch_size: batch size for all batching functions
+    """
+    seqs_batched, seq_lens = length_batch_input_outputs(sequences)
+    len_batching_funcs = {
+        sl: get_batching_func(seq_batch, batch_size)
+        for (sl, seq_batch) in zip(seq_lens, seqs_batched)
+    }
+
+    return len_batching_funcs, seqs_batched, seq_lens
+
+
 def fit(
-    params: Dict,
-    sequences: Set[str],
+    sequences: Iterable[str],
     n_epochs: int,
-    batch_method: Optional[str] = "random",
-    batch_size: Optional[int] = 25,
+    model_func: Callable = mlstm1900_apply_fun,
+    params: Any = load_params(),
```

Comment:
Would it be better to set default to `None` and conditionally call on `load_params` below, such that users can still explicitly set `params=None`. I think I'd be better for code readability.

Reply:
I think I see where you're coming from with this one. Makes sense, the pattern you're suggesting:

```python
fitted_params = fit(..., params=None,...)
```

makes it possible for an end-user to _explicitly_ indicate that no params have been passed in. OK, let me change.